text/0000-reserve-element.md

```diff
@@ -0,0 +1,55 @@
+- Start Date: 2015-12-03
+- RFC PR: (leave this empty)
+- Refract Issue: (leave this empty)
+
+# Summary
+
+This proposes we reserve the keyword `element` for all objects in a Refract
+tree.
+
+# Motivation
+
+This allows us to not worry about domain-specific knowledge when parsing a
+Refract tree. Right now, we have to rely on something out of band to tell us
+when something MAY be refracted, but if we rely on the reserved keyword
+`element`, we can know if an object is a normal object or a Refract Object.
+
+Note: we are already doing this in Minim.
+
+# Design
+
+This RFC proposes to reserve `element` in the entire tree. This means that
+anywhere the `element` keyword is found, a parser SHOULD treat the object as a
+Refract Object.
+
+## Secondary Proposal
+
+This allows for us to know when an object is a Refract Object or not. It also
+allows for us only refract things that need to be refracted when serializing.
+In this situation, for example, the `key` property of a `Member Element` would
+not need to refracted if it is a string.
```

Comment:
So in thinking of this, I don't think we should pursue this too far. But it opens to lots of possibilities. Take this for example:

```js
{
  foo: 'bar',
  baz: 'bab'
}
```

This _could_ be considered valid Refract if we reserve `element`. Now what if the "bar" value is some non-primitive element?

```js
{
  foo: {
    element: 'non-primitive',
    content: 'bar'
  },
  baz: 'bab'
}
```

The only reason we would need to refract the entire object and use member elements here is if `foo` or `baz` have non-primitive elements or have meta or attributes.

---

My concern here is interop with current tooling. But in allowing us to optionally Refract things, it does make things a lot cleaner :)

---

> we must refract member elements

Given the example above, I'm not sure we have to, though interop with cooling tooling could require it. We would only need member elements if (a) the member element has meta or attributes or (b) a key in the object is non-primitive or has meta or attributes. Do you think this would still require a "MUST refract" here?